File: training/python/de/sots/cellarsandcentaurs/domain/model/hit_points.py
from dataclasses import dataclass

MAX_HIT_POINTS = 200


@dataclass(frozen=True)
class HitPoints:
    """Hit points drop when the creature takes damage and recover when it rests in its lair."""

    current: int
    max: int
    temporary: int = 0

    @classmethod
    def init(cls, max_value: int) -> "HitPoints":
        return cls(min(max_value, MAX_HIT_POINTS), min(max_value, MAX_HIT_POINTS), 0)

    def take_damage(self, damage: int) -> "HitPoints":
        return HitPoints(max(self.current - damage, 0), self.max, self.temporary)

File: training/typescript/src/de/sots/cellarsandcentaurs/domain/model/HitPoints.ts
/**
 * Hit points drop when the creature takes damage and recover when it rests in its lair.
 */
export class HitPoints {
    static readonly MAX_HIT_POINTS = 999;

    constructor(
        public current: number,
        public max: number,
        public temporary: number = 0,
    ) {}

    static init(max: number): HitPoints {
        const cappedMax = Math.min(max, HitPoints.MAX_HIT_POINTS);
        return new HitPoints(cappedMax, cappedMax, 0);
    }

    takeDamage(amount: number): void {
        this.current = Math.max(0, this.current - amount);
    }

    rest(): void {
        this.current = this.max;
    }
}
